### PR TITLE
OCPBUGS-2512: Improve IP address sort order for interface selection

### PR DESCRIPTION
**Pull request by andreaskaris**
How to test:
~~~
launch openshift/baremetal-runtimecfg#199 metal,ipv6
~~~

### CODE REVIEW

File: pkg/utils/addresses.go [hunk 1]
```diff
@@ -230,20 +248,32 @@ func addressesDefaultInternal(preferIPv6 bool, af AddressFilter, getAddrs addres
...
 	}
 
-	// Sort addresses into a stable order, based on default route priority and link
+	// Sort addresses into a stable order, based on:
+	// a) default route priority
+	// b) link
+	// c) IP address family
+	// d) The default route matching the IP CIDR
+	// e) The address being public
 	// index. Otherwise the order of the addresses we return may change if an address
```
Comment (cybertron) [inline pkg/utils/addresses.go:265]: Nit: The "index." here is a leftover from the now-removed comment.

File: pkg/utils/addresses.go [hunk 2]
```diff
@@ -94,6 +101,16 @@ func ValidNodeAddress(address netlink.Addr) bool {
 		return false
 	}
 
+	// Ignore the IPv6 OVNK masquerade subnet. We should never see an error here as this is a constant.
+	// If we get an error, panic.
+	_, netip, err := net.ParseCIDR(V6MasqueradeSubnet)
+	if err != nil {
+		panic(err)
+	}
+	if netip.Contains(address.IP) {
```
Comment (cybertron) [inline pkg/utils/addresses.go:110]: I still don't think we can do this unconditionally. The masquerade subnet is a valid address range in non-OVNK CNIs.

Comment (andreaskaris) [reply pkg/utils/addresses.go]: I removed it altogether for this PR

File: pkg/utils/addresses.go (hunk 1 above)
Comment (andreaskaris) [reply pkg/utils/addresses.go]: Changed + renamed one of the variables

File: pkg/utils/addresses.go [hunk 3]
```diff
@@ -230,20 +231,30 @@ func addressesDefaultInternal(preferIPv6 bool, af AddressFilter, getAddrs addres
...
 	}
 
-	// Sort addresses into a stable order, based on default route priority and link
-	// index. Otherwise the order of the addresses we return may change if an address
-	// moves to a bridge (for example).
+	// Sort addresses into a stable order, based on:
+	// a) default route priority
+	// b) link
+	// c) IP address family
+	// d) The gatway being on the IP's CIDR
```
Comment (tsorya) [inline pkg/utils/addresses.go:246]: nit: gateway

**Comment by tsorya** [comment]
lgtm as for me

File: pkg/utils/addresses.go (hunk 3 above)
Comment (andreaskaris) [reply pkg/utils/addresses.go]: fixed, me--

**Comment by cybertron** [comment]
This lgtm. Do we want anyone else to sign off before we merge?

**Comment by andreaskaris** [comment]
@jcaamano

**Comment by andreaskaris** [comment]
> /lgtm
> 
> I am surprised though with the prioritization by link index that was already there. What's that supposed to represent?

From the old comment (which I removed):
~~~
// Sort addresses into a stable order, based on default route priority and link
	// index. Otherwise the order of the addresses we return may change if an address
	// moves to a bridge (for example).
~~~

**Comment by andreaskaris** [comment]
@cybertron I think this is ready to merge, then!

**Comment by cybertron** [comment]
The link index sorting comes from https://github.com/openshift/baremetal-runtimecfg/pull/130 I think that may have actually been unrelated to the address moving to a bridge, but it got included in the same comment because that's what was there.

**Pull request merged**